File: miscellaneous/quick_sort.cpp
#include<bits/stdc++.h>
using namespace std;
typedef long long ll;

int partition(int arr[],int first,int last)
{
    int pivot=arr[last];
    int i=first-1,j;
    for(j=first;j<last;j++)
    {
        if(arr[j]<pivot)
        {
            swap(arr[i+1],arr[j]);
            {
                i++;
            }
        }
    }
    swap(arr[i],pivot);
    return i+1;
}
void quicksort(int arr[],int s,int e)
{
    if(s<e)
    {
        int pivot=partition(arr,s,e);
        quicksort(arr,s,pivot-1);
        quicksort(arr,pivot+1,e);
    }
}
int main()
{
    int arr[]={1,4,8,2,5};
    quicksort(arr,0,4);
    for(int i=0;i<4;i++)
    {
        cout<<arr[i]<<" ";
    }

}

File: Array-DS/quick-sort.py
def partition(arr,l,r):
    pivot=arr[r]
    separator=l-1
    for j in range(l,r+1):
        if arr[j]<=pivot:
            separator=separator+1
            # swap in python
            arr[j],arr[separator]=arr[separator],arr[j]
    return separator

def quicksort(arr,l,r):
    if l<=r:
        pos=partition(arr,l,r)
        quicksort(arr,l,pos-1)
        quicksort(arr,pos+1,r)
        return arr

arr=[10,80,30,90,40,50,70]

print(quicksort(arr,0,len(arr)-1))
